src/mlpack/methods/lsh/lsh_search_impl.hpp

```diff
@@ -331,12 +333,423 @@ void LSHSearch<SortPolicy>::BaseCase(const size_t queryIndex,
+
+  if ( maxPos + 1 < A.size()) // otherwise, this is an invalid vector
+  {
+    A[maxPos + 1] = 1;
+    return true;
+  }
+  return false;
+
+}
+
+// Return true if perturbation set A is valid. A perturbation set is invalid if
+// it contains two (or more) actions for the same dimension or dimensions that
+// are larger than the queryCode's dimensions.
+inline bool perturbationValid(const std::vector<bool>& A,
+                              const size_t numProj)
+{
+  // Use check to mark dimensions we have seen before in A. If a dimension is
+  // seen twice (or more), A is not a valid perturbation.
+  std::vector<bool> check(numProj);
+
+  if (A.size() > 2 * numProj)
+    return false; // This should never happen.
+
+  // Check that we only see each dimension once. If not, vector is not valid.
+  for (size_t i = 0; i < A.size(); ++i)
+  {
+    // Only check dimensions that were included.
+    if (!A[i])
+      continue;
+
+    // If dimesnion is unseen thus far, mark it as seen.
+    if ( check[i % numProj] == false )
+      check[i % numProj] = true;
+    else
+      return false; // If dimension was seen before, set is not valid.
+  }
+
+  // If we didn't fail, set is valid.
+  return true;
+}
+
+// Compute additional probing bins for a query
+template<typename SortPolicy>
+void LSHSearch<SortPolicy>::GetAdditionalProbingBins(
+                            const arma::vec& queryCode,
+                            const arma::vec& queryCodeNotFloored,
+                            const size_t T,
+                            arma::mat& additionalProbingBins) const
+{
+
+  // No additional bins requested. Our work is done.
+  if (T == 0)
+    return;
+
+  // Each column of additionalProbingBins is the code of a bin.
+  additionalProbingBins.set_size(numProj, T);
+  
+  // Copy the query's code, then in the end we will  add/subtract according 
+  // to perturbations we calculated.
+  for (size_t c = 0; c < T; ++c)
+    additionalProbingBins.col(c) = queryCode;
+
+
+  // Calculate query point's projection position.
+  arma::mat projection = queryCodeNotFloored;
+
+  // Use projection to calculate query's distance from hash limits.
+  arma::vec limLow = projection - queryCode * hashWidth;
+  arma::vec limHigh = hashWidth - limLow;
+
+  // Calculate scores. score = distance^2.
+  arma::vec scores(2 * numProj);
+  scores.rows(0, numProj - 1) = arma::pow(limLow, 2);
+  scores.rows(numProj, (2 * numProj) - 1) = arma::pow(limHigh, 2);
+
+  // Actions vector describes what perturbation (-1/+1) corresponds to a score.
+  arma::Col<short int> actions(2 * numProj); // will be [-1 ... 1 ...]
+  actions.rows(0, numProj - 1) = // first numProj rows
+    -1 * arma::ones< arma::Col<short int> > (numProj); // -1s
+  actions.rows(numProj, (2 * numProj) - 1) = // last numProj rows
+    arma::ones< arma::Col<short int> > (numProj); // 1s
+
+
+  // Acting dimension vector shows which coordinate to transform according to
+  // actions described by actions vector
```

Comment:
I always feel so pedantic writing this, but can you finish the sentence with a period?  I think it's important to have complete sentences in documentation for clarity, so that people for whom English isn't the first language can have a better understanding.